GCEWindowsAgent/oslogin.go

```diff
@@ -0,0 +1,329 @@
+//  Copyright 2019 Google Inc. All Rights Reserved.
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+
+package main
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"runtime"
+	"strings"
+
+	"github.com/GoogleCloudPlatform/guest-logging-go/logger"
+)
+
+var (
+	googleComment    = "# Added by Google Compute Engine OS Login."
+	googleBlockStart = "#### Google OS Login control. Do not edit this section. ####"
+	googleBlockEnd   = "#### End Google OS Login control section. ####"
+)
+
+type osloginMgr struct{}
+
+func (o *osloginMgr) diff() bool {
+	// True on first run.
+	return oldMetadata.Project.ProjectID == "" ||
+		// True if any value has changed.
+		(oldMetadata.Instance.Attributes.EnableOSLogin != newMetadata.Instance.Attributes.EnableOSLogin) ||
+		(oldMetadata.Instance.Attributes.TwoFactor != newMetadata.Instance.Attributes.TwoFactor) ||
+		(oldMetadata.Project.Attributes.EnableOSLogin != newMetadata.Project.Attributes.EnableOSLogin) ||
+		(oldMetadata.Project.Attributes.TwoFactor != newMetadata.Project.Attributes.TwoFactor)
+}
+
+func (o *osloginMgr) timeout() bool {
+	return false
+}
+
+func (o *osloginMgr) disabled(os string) bool {
+	return os == "windows"
+}
+
+func (o *osloginMgr) set() error {
+	oldEnable := oldMetadata.Instance.Attributes.EnableOSLogin || oldMetadata.Project.Attributes.EnableOSLogin
+	enable := newMetadata.Instance.Attributes.EnableOSLogin || newMetadata.Project.Attributes.EnableOSLogin
+	twofactor := newMetadata.Instance.Attributes.TwoFactor || newMetadata.Project.Attributes.TwoFactor
+
+	if enable && !oldEnable {
+		logger.Infof("Enabling OS Login")
+		newMetadata.Instance.Attributes.SSHKeys = nil
+		newMetadata.Project.Attributes.SSHKeys = nil
+		(&linuxAccountsMgr{}).set()
+	}
+
+	if err := updateSSHConfig(enable, twofactor); err != nil {
+		logger.Errorf("Error updating SSH config: %v.\n", err)
```

Comment:
Ah, I missed this before, log lines inherently end in a new line correct? Wouldn't this add an extra one?